rustls/src/msgs/deframer.rs

```diff
@@ -87,9 +91,20 @@ impl MessageDeframer {
                 }
             };
 
-            // Return CCS messages immediately without decrypting.
+            // Return CCS messages and plaintext alerts received before any encrypted messages
+            // immediately without decrypting.
+            //
+            // * CCS messages are always plaintext.
+            // * Alerts in TLS 1.3 _may_ be received plaintext early in the handshake process,
+            //   but are not expected once encrypted data has been exchanged (e.g. read_seq >0).
+            //   We also expect all plaintext alerts to be at most 2 bytes in size.
             let end = start + rd.used();
-            if m.typ == ContentType::ChangeCipherSpec && self.joining_hs.is_none() {
+            let type_is_ccs = m.typ == ContentType::ChangeCipherSpec;
+            let early_plaintext_alert = m.typ == ContentType::Alert
+                && m.payload.0.len() <= 2
+                && record_layer.read_seq() == 0
```

Comment:
I'm afraid I don't think this is right, because `read_seq` can be reset to 0 later on every receipt of a `KeyUpdate`. Instead, I think we should expose `pub(crate) fn is_decrypting(&self) -> bool` (like the existing `is_encrypting`)

Reply:
@ctz I don't think this will work. When reproducing the condition from #1369 the `RecordLayer` has `encrypt_state = DirectionState::Active` and `decrypt_state = DirectionState::Active`, even though no encrypted message has been received by the peer - both the `write_seq` and `read_seq` are 0.

Reply:
> Did you figure out when/where the decrypt_state gets set to Active?

Yeah, the TLS 1.3 `ExpectClientHello` state `handle_client_hello` with early data disabled is setting the handshake decrypter, which moves the record layer from `DirectionState::Invalid` -> `DirectionState::Prepared` -> `DirectionState::Active` in one operation via the `KeySchedule` calling `RecordLayer.set_message_decrypter`.

We could consider another `DirectionState` variant between `Prepared` and `Active` but it feels a little clumsy :thinking:

Reply:
What about having the `RecordLayer` keep a `bool` for whether it has decrypted a message separate from the `read_seq`?

The `read_seq` would get reset on when processing a `KeyUpdate` but not the proposed `bool`.

Reply:
> What about having the `RecordLayer` keep a `bool` for whether it has decrypted a message separate from the `read_seq`?
> 
> The `read_seq` would get reset when processing a `KeyUpdate` but not the proposed `bool`.

Yes I think that ought to work and seems like the least bad option.

Reply:
Implemented in https://github.com/rustls/rustls/pull/1392/commits/9627b16d00f57dd2e7d849ec57e5f2baa0bef9fc